[Catalina GPU] fast/canvas/webgl/lose-context-on-timeout.html is a flaky timeout

djg opened the pull request:
#### 4a252fa8dc083a715095358068c85d32bc99b33a
<pre>
[Catalina GPU] fast/canvas/webgl/lose-context-on-timeout.html is a flaky timeout
<a href="https://bugs.webkit.org/show_bug.cgi?id=229523">https://bugs.webkit.org/show_bug.cgi?id=229523</a>
&lt;rdar://problem/82360416 &gt;

Fix the race conditions in simulating GPUP GraphicsContextGL timeout by
switching from async to sync calls.  This ensures that the processing of
subsequent messages will fail, triggering webglcontextlost event which the test
relies on to complete successfully.

Reviewed by NOBODY (OOPS!).

* Tools/Scripts/generate-gpup-webgl
- Add missing LIBGBM PrepareForDisplay()
- Make SimulateEventForTesting(...) synchronous so subsequent
* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.cpp:
(WebKit::GPUConnectionToWebProcess::dispatchMessage):
(WebKit::GPUConnectionToWebProcess::dispatchSyncMessage):
GraphicsContextGL messages received out-of-stream are invalid and + should be
treated as a failure.
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.cpp:
(WebKit::RemoteGraphicsContextGL::workQueueUninitialize):
Don&apos;t unconditionally access m_context. In the testing case of
lose-context-on-timeout, m_context is NULL on shutdown.
(WebKit::RemoteGraphicsContextGL::simulateEventForTesting):
Make SimulatedEventForTesting::Timeout case synchronous. Making the IPC message
SimulateEventForTesting synchronous and using callOnMainRunLoopAndWait fixes the
flaky behavior of lose-context-on-timeout.html
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.h:
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.messages.in:
Source changes due to generate-gpup-webgl
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
Add generated files for RemoteGraphicsContextGL and RemoteGraphicsContextGLProxy.
* Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextGLProxy.cpp:
(WebKit::RemoteGraphicsContextGLProxy::simulateEventForTesting):
* LayoutTests/fast/canvas/webgl/lose-context-on-timeout-expected.txt:
Correct output from test.
* LayoutTests/fast/canvas/webgl/lose-context-on-timeout.html:
Add a second WebGL context to keep the WP-GPUP connection alive. This stops the
WP-GPUP connection from being closed when no more GraphicsContextGL remain,
causing the test to pass.
* LayoutTests/gpu-process/TestExpectations:
* LayoutTests/platform/mac-wk2/TestExpectations:
GPUP is required for correct execution of lose-context-on-timeout.html.
</pre>

kkinnunen-apple commented on Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.cpp:
```diff
@@ -139,7 +139,8 @@ void RemoteGraphicsContextGL::workQueueInitialize(WebCore::GraphicsContextGLAttr
-    m_context->setClient(nullptr);
+    if (m_context)
+        m_context->setClient(nullptr);
     m_context = nullptr;
```
This is the only intended place that currently clears the m_context, is there's something strange going on if it can already be nullptr. Could you investigate that still?

djg replied:
This was left over from another change. Removed.

kkinnunen-apple commented on Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.cpp:
```diff
@@ -284,29 +285,33 @@ void RemoteGraphicsContextGL::copyTextureFromVideoFrame(WebKit::RemoteVideoFrame
+        break;
     }
-    m_context->simulateEventForTesting(event);
+    completionHandler();
```
Yeah, this is a nice simple solution to the problem.. it may create problems down the line, but let's go with this one for now..

Technically it can still race executing the commands in the buffer. 
- GPUP processes the message buffer. It sees simulateEventForTesting.
- GPUP executes simulateEventForTesting. This removes the stream connection from the work queue. This does not stop the message buffer processing.
- WP sees simulateEventForTesting reply. WP submits the 3 commands after simulateEventForTesting finishes.
- GPUP still processes the message buffer and sees new commands and processes those and their replies.

I think we'll need to implement `StreamServerConnection::invalidate()` that immediately stops the buffer processing. However, that creates the problem "down the line", where the reply by `completionHandler()` will not be delivered at that point. But maybe when that's implemented, we just switch this message back to async.

djg replied:
In local testing, I'm still seeing the test timeout even after making all the calls sync.

Back to the drawing board.

djg committed:
[Catalina GPU] fast/canvas/webgl/lose-context-on-timeout.html is a flaky timeout
https://bugs.webkit.org/show_bug.cgi?id=229523
<rdar://problem/82360416>

Fix the race conditions in simulating GPUP GraphicsContextGL timeout by
switching from async to sync calls.  This ensures that the processing of
subsequent messages will fail, triggering webglcontextlost event which the test
relies on to complete successfully.

Reviewed by NOBODY (OOPS!).

* Tools/Scripts/generate-gpup-webgl
- Add missing LIBGBM PrepareForDisplay()
- Make SimulateEventForTesting(...) synchronous so subsequent
* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.cpp:
(WebKit::GPUConnectionToWebProcess::dispatchMessage):
(WebKit::GPUConnectionToWebProcess::dispatchSyncMessage):
GraphicsContextGL messages received out-of-stream are invalid and + should be
treated as a failure.
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.cpp:
(WebKit::RemoteGraphicsContextGL::workQueueUninitialize):
Don't unconditionally access m_context. In the testing case of
lose-context-on-timeout, m_context is NULL on shutdown.
(WebKit::RemoteGraphicsContextGL::simulateEventForTesting):
Make SimulatedEventForTesting::Timeout case synchronous. Making the IPC message
SimulateEventForTesting synchronous and using callOnMainRunLoopAndWait fixes the
flaky behavior of lose-context-on-timeout.html
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.h:
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.messages.in:
Source changes due to generate-gpup-webgl
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
Add generated files for RemoteGraphicsContextGL and RemoteGraphicsContextGLProxy.
* Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextGLProxy.cpp:
(WebKit::RemoteGraphicsContextGLProxy::simulateEventForTesting):
* LayoutTests/fast/canvas/webgl/lose-context-on-timeout-expected.txt:
Correct output from test.
* LayoutTests/fast/canvas/webgl/lose-context-on-timeout.html:
Add a second WebGL context to keep the WP-GPUP connection alive. This stops the
WP-GPUP connection from being closed when no more GraphicsContextGL remain,
causing the test to pass.
* LayoutTests/gpu-process/TestExpectations:
* LayoutTests/platform/mac-wk2/TestExpectations:
GPUP is required for correct execution of lose-context-on-timeout.html.
